docker/prepare_proxy.sh

```diff
@@ -40,20 +51,53 @@ if [[ -z "${ISTIO_PROXY_PORT-}" ]] || [[ -z "${ISTIO_PROXY_UID-}" ]]; then
   exit 1
 fi
 
-# 1. External traffic is redirected to proxy port
-iptables -t nat -A PREROUTING -p tcp -j REDIRECT --to-port ${ISTIO_PROXY_PORT}
-
-# 2. Locally generated traffic sent on loopback interface (because traffic was
-# routed locally) and destination IP is not explicitly the loopback IP
-# (e.g. 172.17.0.2 and not 127.0.0.1) is redirected back to proxy port.
-# This include any proxy generated traffic which is necessary to make
-# (app => proxy => proxy => app) work when two apps are same pod.
-iptables -t nat -A OUTPUT -p tcp -o lo -j REDIRECT \
-  ! -d 127.0.0.1/32 --to-port ${ISTIO_PROXY_PORT}
-
-# 3. Locally generated traffic not sent explicitly from loopback IP
-# (i.e. proxy aware) or not from the proxy itself is redirected proxy port.
-iptables -t nat -A OUTPUT -p tcp -j REDIRECT ! -s 127.0.0.1/32 \
-  --to-port ${ISTIO_PROXY_PORT} -m owner '!' --uid-owner ${ISTIO_PROXY_UID}
+iptables -t nat -F
+
+# Create a new chain for redirecting inbound and outbound traffic to
+# the common Istio proxy port.
+iptables -t nat -N ISTIO_REDIRECT                                                   -m comment --comment "istio/redirect-common-chain"
+iptables -t nat -A ISTIO_REDIRECT -p tcp -j REDIRECT --to-ports ${ISTIO_PROXY_PORT} -m comment --comment "istio/redirect-to-proxy-port"
+
+# Redirect all inbound traffic to the proxy.
+iptables -t nat -A PREROUTING -j ISTIO_REDIRECT                                     -m comment --comment "istio/install-istio-prerouting"
+
+# Create a new chain for selectively redirecting outbound packets to
+# istio and attach it the OUTPUT rule for all tcp traffic. '-j RETURN'
+# bypasses the proxy port; '-j ISTIO_REDIRECT' redirects to the proxy
+# port.
+iptables -t nat -N ISTIO_OUTPUT
+iptables -t nat -A OUTPUT -p tcp -j ISTIO_OUTPUT                                    -m comment --comment "istio/install-istio-output"
+
+# Locally routed traffic is not captured by PREROUTING chain. Redirect
+# loopback back traffic when the DST_IP is not explicitly the loopback
+# address to handles appN => proxy (client) => proxy (server) => appN.
+iptables -t nat -A ISTIO_OUTPUT -o lo ! -d 127.0.0.1/32 -j ISTIO_REDIRECT           -m comment --comment "istio/implicit-loopback"
```

Comment:
Probably best to describe the use case instead of the data flow. E.g., when kubelet accesses the app for health checks (??), or app calls itself via http://podIP/..